ktxspec.adoc

```diff
@@ -1030,17 +1032,16 @@ The Universal ASTC image format (UASTC) is indicated by `colorModel`
 of <<KDF13>>. Images in this format must be transcoded to a
 GPU-supported block-compressed format or decoded to a GPU-supported
 uncompressed format before being uploaded to and sampled by a GPU.
-UASTC images can be supercompressed with Zstandard
-(`supercompressionScheme` = 2) with or without first conditioning
-the data with Rate-distortion Optimization. If supercompression is
-used, the DFD's `bytesPlane[0-7]` must be set to 0 as described in
-the next subsection.
+UASTC images can be supercompressed with any scheme except BasisLZ
+(`supercompressionScheme` = 1). If supercompression is used, the
+DFD's `bytesPlane[0-7]` must be set to 0 as described in the next
+subsection.
```

Comment:
Although existed for quite some time, this language and the later subsection about DFD for supercompressed formats are fundamentally incorrect for UASTC. The `bytesPlane[0-7]` fields set to 0 represent an unsized format but UASTC and UASTC HDR are sized unlike ETC1S/BasisLZ.

Additionally, see #217;  Zstandard and Deflate suprecompression schemes should not affect DFD at all.

Reply:
? I thought setting "bytesPlane" to 0 was a way of saying "this has not been decompressed into the block format". The non-supercompressed block representation for UASTC/UASTC HDR should have a coherent bytesPlane value, but how does having a fixed bytesPlane value make sense if the data is supercompressed?

Reply:
The term "supercompression" is a bit overloaded.

The BasisLZ scheme is inherently not block-based, i.e., it's decoded to a set of in-memory structures that are later used to generate target GPU formats. There is no point in the decoding pipeline where a literal ETC1 block layout is used except when it's the requested output. This allows us to say that BasisLZ/ETC1S data is _unsized_ in DFD-speak.

For comparison, Zstandard and Deflate schemes are common-purpose lossless compression algorithms that could be applied on top of _any_ pixel format. These schemes are 100% transparent to the decoding pipeline and should be seen as a separate feature layer that does not affect the DFD.

Reply:
Ah, since deflate isn't a texture-like format my assumption would be that it, too, doesn't have a standard "block size". But I defer to those who deal with it!

Reply:
The main reason to not lose `bytesPlane` when Deflate is used is to still be able to understand data layout without depending on `vkFormat`. If the data is unrepresentable with a Vulkan format, then applying Deflate and resetting all `bytesPlane` values to zeros would make the file unreadable.

Reply:
> ? I thought setting "bytesPlane" to 0 was a way of saying "this has not been decompressed into the block format". 

It is.

> The BasisLZ scheme is inherently not block-based,

BasisLZ only appears different due to its implementation which wraps ETC1S encoding and LZ-style compression together. Conceptually it is no different than, e.g. UASTC + Zstd - a block-compressed format deflated with a scheme that produces variable length output. As noted above bytesPlane0 is used to indicate that there is not a randomly-accessible texel layout in memory and decoding to that layout is necessary before the DFD can be used to inspect texels/blocks.

> The main reason to not lose bytesPlane when Deflate is used is to still be able to understand data layout without depending on vkFormat

There is existence proof [here](https://github.com/KhronosGroup/KTX-Software/blob/8aa500a1f0a59abe22d4ef5e6842d563510dd594/external/dfdutils/queries.c#L115) of reconstructing bytesPlane without use of vkFormat and a [unittest](https://github.com/KhronosGroup/KTX-Software/blob/8aa500a1f0a59abe22d4ef5e6842d563510dd594/tests/unittests/unittests.cc#L693) that tests it for the DFD created for every VkFormat. (I should probably add tests for the DFDs for the Basis Universal formats.) I see no reason to change the current specification.

Reply:
Thanks for the references.

That function's source code highlights the main issue: sample's `bitOffset` and `bitLength` cannot represent trailing padding within a pixel, hence the hard-coded special cases for depth and stencil formats. Admittedly, I missed that detail when the initial supercompression support was specified for KTX 2.0 but it's not a reason to keep it this way.

Another practical example would be "RGBX8": three 8-bit color channels followed by 8 unused bits. Since DFD has no notion of "empty" or "padding" samples, setting `bytesPlane0` to 4 is the only way to represent this layout. Note that Section 5.12 of KDFS explicitly mentions that:
> There is no requirement that samples should access unique parts of the bit-stream: formats such as combined intensity and alpha, or shared exponent formats, require that bits be reused. Nor is there a requirement that all the bits in a plane be used (a format may contain padding).

Besides, it does not make much sense to tie DFD to supercompression in the first place. DFD describes per-pixel layout while Deflate and Zstandard operate on whole levels without looking into pixel rows let alone samples. Therefore, compressing, e.g., RGBA8 (or RGBX8) should change only per-level offsets/lengths in the KTX header.

Reply:
The function correctly handles all these formats that have padding (and any others in the format list I may have missed in my quick scan):

```
VK_FORMAT_R10X6_UNORM_PACK16,
VK_FORMAT_R10X6G10X6_UNORM_2PACK16,
VK_FORMAT_R10X6G10X6B10X6A10X6_UNORM_4PACK16,
VK_FORMAT_G10X6B10X6G10X6R10X6_422_UNORM_4PACK16,
VK_FORMAT_B10X6G10X6R10X6G10X6_422_UNORM_4PACK16,
VK_FORMAT_R12X4_UNORM_PACK16,
VK_FORMAT_R12X4G12X4_UNORM_2PACK16,
VK_FORMAT_R12X4G12X4B12X4A12X4_UNORM_4PACK16,
VK_FORMAT_G12X4B12X4G12X4R12X4_422_UNORM_4PACK16,
VK_FORMAT_B12X4G12X4R12X4G12X4_422_UNORM_4PACK16,
```

IIRC this is because the padding is in the low order bits so taking (largest bitOffset) + bitLength + 1 gives the total number of bits in the texel.

Maybe this would break if the DFD was modified to describe the data laid out on a big-endian machine. I don't know. Thinking about a DFD for big-endian data makes my head spin. @fluppeteer?

The reason for the special cases for depth and stencil is mostly because it is simpler. They can be identified by the channelId and as a result there is no need to loop over the samples.

> Besides, it does not make much sense to tie DFD to supercompression in the first place. DFD describes per-pixel layout while Deflate and Zstandard operate on whole levels without looking into pixel rows let alone samples.

The point is that there _is no_ per-pixel layout when the data is supercompressed. and the user needs to know that. While KTX files have the supercompression field to tell users the data is supercompressed, other uses of a DFD may not. I think it is helpful that the DFD spec. has this requirement so users know they can't just go to the address of the data and start interpreting it via the DFD.

Reply:
> The function correctly handles all these formats that have padding ...

All those formats happen to define _bottom_ bits as unused so `bitOffset + bitLength + 1` works. This assumption does not work for any format (current or future) that uses _top_ bits for padding.

> The reason for the special cases for depth and stencil is mostly because it is simpler. They can be identified by the channelId and as a result there is no need to loop over the samples.

Those special cases exist only because those three formats use top bits for padding (see the comment in the source code):
- `X8_D24_UNORM_PACK32` puts data in bottom bits so its naively computed `bytesPlane0` would be 3;
- `D16_UNORM_S8_UINT` (as defined in KTX) puts stencil data in bottom bits so its naively computed `bytesPlane0` would also be 3;
- `D32_SFLOAT_S8_UINT` (as defined in KTX) puts stencil data in bottom bits so its naively computed `bytesPlane0` would be 5.

The mere existence of these special cases proves that `bytesPlane0` is not redundant and therefore cannot be omitted. For example, DFD wouldn't be able to differentiate between padded (4-byte) and tight (3-byte) D24 layouts otherwise.

> The point is that there _is no_ per-pixel layout when the data is supercompressed. and the user needs to know that. While KTX files have the supercompression field to tell users the data is supercompressed, other uses of a DFD may not.

Sorry, I don't follow. Even when `supercompressionScheme` zero, users still have to read at least all dimensions including level and face counts before making any use of DFD. In practice, they will also rely on the level index instead of computing every offset manually. With these prerequisites in place, there is absolutely no chance that someone would try to use DFD embedded in a KTX file to read the data from the same KTX file without looking at other KTX header fields.

Reply:
> All those formats happen to define bottom bits as unused so bitOffset + bitLength + 1 works. This assumption does not work for any format (current or future) that uses top bits for padding.

Yes. I said the same as the first sentence in my previous message and implied what is in the second sentence.

> > The point is that there is no per-pixel layout when the data is supercompressed. ...

What I mean is that while supercompressed the data is not as described by the DFD.

> Sorry, I don't follow. Even when the supercompressionScheme is zero, users still have to read at least all dimensions including level and face counts ...

I was talking about uses of a DFD outside of KTX and not necessarily for textures. The requirement for bytesPlane = 0 comes from KDFS not KTX. If we change it we have to change KDFS as well. I agree there could be rare formats for which the reconstruction will not work. Perhaps KDFS needs to find another way to indicate the data is supercompressed so we can keep bytesPlane0 as is. @fluppeteer any thoughts?

Even if we change it, libktx will have to support loading KTX files with bytesPlane = 0 for the foreseeable future.

Reply:
> The requirement for bytesPlane = 0 comes from KDFS not KTX.

The "unsized" formats in KDFS are rather loosely defined and the relevant language certainly does not imply that any use of compression anywhere in the image pipeline _must_ reset `bytesPlane*` values to zeros. That's given only as a hypothetical example. I agree that it should be better worded.

KTX can easily avoid any ambiguities in this area by saying that DFD describes pixel data after supercompression has been decoded (with an extra note regarding BasisLZ). `bytesPlane*` values would be the only observable difference compared to the current situation.

>  libktx will have to support loading KTX files with bytesPlane = 0 for the foreseeable future

Yes, the upgrade path should be implemented carefully.

Reply:
I've opened https://github.com/KhronosGroup/DataFormat/issues/27 requesting the change to KDFS. Further discussion should take place over there. I'm going to mark this issue as resolved here since it is tangential to the topic of this PR.